DEV: Pass additional runtests.py args to ASV

seiko2plus opened the pull request:
#### Pass all extra arguments that start with `--` to ASV 

Examples:
        `runtests.py` `--bench-compare master sin --cpu-affinity 1`
        `runtests.py` `--bench-compare master -- --cpu-affinity 1`
        `runtests.py` `--bench sin --cpu-affinity 1`
        `runtests.py` `--bench -- --cpu-affinity 1`

seberg commented:
Seems like this can be useful on its own, although I wonder if fixing `--` together with `--bench` is not better. Maybe (although its not perfect probably):
```diff
diff --git a/runtests.py b/runtests.py
index e470f8a9d..2d9a0f39d 100755
--- a/runtests.py
+++ b/runtests.py
@@ -220,7 +220,14 @@ def main(argv):
 
     if args.bench:
         # Run ASV
-        items = extra_argv
+        if "--" in extra_argv:
+            split = extra_argv.index("--")
+            items = extra_argv[:split]
+            extra_argv = extra_argv[split+1:]
+        else:
+            items = extra_argv
+            extra_argv = []
+
         if args.tests:
             items += args.tests
         if args.submodule:
@@ -229,6 +236,7 @@ def main(argv):
         bench_args = []
         for a in items:
             bench_args.extend(['--bench', a])
+        bench_args.extend(extra_argv)
 
         if not args.bench_compare:
             cmd = ['asv', 'run', '-n', '-e', '--python=same'] + bench_args
```
plus maybe as an example in the initial help (it also doesn't list asv for `--`).

seberg commented:
@seiko2plus if you think changing the script to pass `--` is a good idea, do you want to do that? I assume nobody will mind, and since its an internal script, we can just finish this off quickly.

seiko2plus committed:
DEV: Pass additional runtests.py args to ASV

Since arguments list benchmarks, all arguments after the first
unknown `--` starting (removing a single first `--`) are passed
on to asv. Previously this was the right pattern for passing
to pytest, so it now works also for ASV.

Examples:
    `runtests.py` `--bench-compare master sin --cpu-affinity 1`
                  `--bench-compare master -- --cpu-affinity 1`
                  `--bench sin --cpu-affinity 1`
                  `--bench -- --cpu-affinity 1`

seiko2plus commented:
@seberg, alright I modified the patch with your proposal, it's more practical I guess. thank u!.

seberg commented:
@seiko2plus sorry, since this is a dev thing, I think we should just do it. I pushed a commit to make things work with `--` (i.e. everything after a `--` is passed on and the `--` iself is skipped, but actual benchmarks must be listed before the `--`). That seemed the most reasonable thing to me, although I am not sure `asv` actually uses any non `--` args, so it may not matter.

seiko2plus commented:
@seberg, your changes make it more robust since it allows passing all tokens after `--` even the ones that start with an only single dash.

seiko2plus commented on runtests.py:
```diff
@@ -127,7 +127,7 @@ def main(argv):
                         help="Arguments to pass to Nose, Python or shell")
     args = parser.parse_args(argv)
-
+    print(args)
```
@seberg, you not gonna keep it, right?

seberg replied:
Ops, no, of course not, thanks :). I was confused about how argparse does things... and I think now it does it "the same way" with `--bench` (at least kind of).

seiko2plus replied:
I do the same most of the time, git hook for detecting such case like this may help u.

seberg replied:
Maybe (some general hooks on NumPy that are easy to enable could be neat too)... anyway thanks @seiko2plus, I will put this in, its easy enough to change if its not yet ideal...

seberg commented:
For everyone else, since this only affects devlopment, and even then only the `--bench` option to `runtests.py`, I will just merge this soon I think.

seiko2plus commented:
@seberg, thank you
